### PR TITLE
[BUGFIX] Filter record metadata value based on metadata property policies

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

This PR fixes metadata visibility errors when fetching records for a user.

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: argilla-server/src/argilla_server/apis/v1/handlers/datasets/records.py [hunk 1]
```diff
@@ -570,8 +569,8 @@ async def search_current_user_dataset_records(
     limit: int = Query(default=LIST_DATASET_RECORDS_LIMIT_DEFAULT, ge=1, le=LIST_DATASET_RECORDS_LIMIT_LE),
     current_user: User = Security(auth.get_current_user),
 ):
-    dataset = await _get_dataset_or_raise(db, dataset_id, with_fields=True)
-
+    dataset = await _get_dataset_or_raise(db, dataset_id, with_fields=True, with_metadata_properties=True)
+    print(dataset.metadata_properties)
```
Comment (frascuchon) [inline argilla-server/src/argilla_server/apis/v1/handlers/datasets/records.py]:
```suggestion
```

File: argilla-server/src/argilla_server/apis/v1/handlers/datasets/records.py [hunk 2]
```diff
@@ -603,6 +602,13 @@ async def search_current_user_dataset_records(
     )
 
     for record in records:
+        record.dataset = dataset
+
+        for metadata_name in list(record.metadata_.keys()) or {}:
+            if not await is_authorized(current_user, RecordPolicyV1.get_metadata(record, metadata_name)):
+                print(f"not authorized for {metadata_name}")
```
Comment (frascuchon) [inline argilla-server/src/argilla_server/apis/v1/handlers/datasets/records.py]:
```suggestion 
```

File: argilla-server/src/argilla_server/policies.py [hunk 3]
```diff
@@ -582,6 +582,24 @@ async def is_allowed(actor: User) -> bool:
...
         return is_allowed
 
+    @classmethod
+    def get_metadata(cls, record: Record, metadata_name: str):
+        async def is_allowed(actor: User) -> bool:
+            if actor.is_owner:
+                return True
+
+            metadata_property = record.dataset.metadata_property_by_name(metadata_name)
+            print(metadata_property, f"for {metadata_name}")
```
Comment (frascuchon) [inline argilla-server/src/argilla_server/policies.py]:
```suggestion
```

File: argilla-server/src/argilla_server/policies.py [hunk 4]
```diff
@@ -582,6 +582,24 @@ async def is_allowed(actor: User) -> bool:
...
+            if metadata_property:
+                print(metadata_property, metadata_property.allowed_roles)
```
Comment (frascuchon) [inline argilla-server/src/argilla_server/policies.py]:
```suggestion
```

**Review by jfcalvo — APPROVED** [review]
My only concern is that this change potentially can add a lot of database queries to check for the metadata. 

Apart from that it looks good to me.

File: argilla-server/src/argilla_server/apis/v1/handlers/datasets/records.py [hunk 5]
```diff
@@ -698,3 +703,15 @@ async def list_dataset_records_search_suggestions_options(
             for sa in suggestion_agents_by_question
         ]
     )
+
+
+async def _filter_record_metadata_for_user(record: Record, user: User) -> Optional[Dict[str, Any]]:
+    print(record.metadata_)
```
Comment (frascuchon) [inline argilla-server/src/argilla_server/apis/v1/handlers/datasets/records.py]:
```suggestion
```

**Pull request merged**